RetroAchievements/RAWeb
# feat: migrate `/forums/recent-posts` to Inertia.js+React

This PR migrates the `/forums/recent-posts` to Inertia.js+React. Rather than delete the PHP code, it still exists. The React route can be accessed at `/forums/recent-posts2`. This is intended to help compare and contrast the existing page vs the rewritten page.

In a follow-up PR, the PHP page can be removed and replaced with the new one.

**How to test this PR:**
* `composer install` and `npm install`
* `sail up`
* `npm run build`
* With Sail running in the background, `sail artisan inertia:start-ssr`
* Navigate to http://localhost:64000/forums/recent-posts2

`sail artisan inertia:start-ssr` causes the React code to render on the server. For development purposes, it should not be used and the conventional `npm run dev` should be used instead.

![Screenshot 2024-06-15 at 2 01 10 PM](https://github.com/RetroAchievements/RAWeb/assets/3984985/5896ba0d-faa2-4ef8-b5d5-5b4f2eac58da)

There should only be two functional changes - quite minor.

**1. Breadcrumbs are slightly different.**
![Screenshot 2024-06-15 at 2 01 46 PM](https://github.com/RetroAchievements/RAWeb/assets/3984985/9cbae93e-ce29-45c8-ab5b-a651d3904f17)

The breadcrumbs are built with [shadcn-ui/Breadcrumb](https://ui.shadcn.com/docs/components/breadcrumb). These are battle-tested and accessible. We can style these however we like and can match the existing style if we so choose, but I think this newer style works much better for both light and dark modes.

**2. Pagination is slightly different.**
![Screenshot 2024-06-15 at 2 03 15 PM](https://github.com/RetroAchievements/RAWeb/assets/3984985/a39875ab-4fb2-496a-b040-e29e35f210b7)

Icons are being used instead of keyboard angle brackets (`<` and `>`).

---

The React components have very high test coverage. `npm run test` has been modified to be similar to `composer test`, using a dot reporter. You can find the tests in the project by searching for any file with the pattern _*.test.tsx_.

---

**Next steps:**
Assuming this PR is approved and merged, in a follow-up PR:
* The PHP code for `/forums/recent-posts` will be deleted.
* The query to build all the props for the page will be migrated to Eloquent ORM and likely called via an action.
* All the temporary code/TODOs in `ForumTopicCommentController::showRecentPosts()` will be removed.
